File: typescript/leetcode/0011-container-with-most-water/solution01.ts
/**
 * @param {number[]} height
 * @return {number}
 */
export function maxArea(height: number[]): number {
  let x1 = 0
  let x2 = height.length - 1
  let maxarea = 0

  while (x1 < x2) {
    const y1 = height[x1]
    const y2 = height[x2]

    const area = (x2 - x1) * Math.min(y1, y2)
    maxarea = Math.max(maxarea, area)

    if (y1 < y2) {
      x1 += 1
    } else {
      x2 -= 1
    }
  }

  return maxarea
}

File: python/leetcode/0011-container-with-most-water/solution01.py
from typing import List

class Solution:
    def maxArea(self, height: List[int]) -> int:
        x1 = 0
        x2 = len(height) - 1
        maxarea = 0

        while x1 < x2:
            y1 = height[x1]
            y2 = height[x2]

            area = min(y1, y2) * (x2 - x1)
            maxarea = max(maxarea, area)

            # I didn't need to find next bigger and closest index
            # since I compare the areas with =max()=
            if y1 < y2:
                shorter = y1
                start = x1
                end = x2
                step = 1
            else:
                shorter = y2
                start = x2
                end = x1
                step = -1

            bigger = None
            for i in range(start, end, step):
                if height[i] > shorter:
                    bigger = i
                    break
            if bigger is None:
                break

            if y1 < y2:
                x1 = bigger
            else:
                x2 = bigger

        return maxarea
